### PR TITLE
Towards Tock 1.0 System Call Interface

### PR DESCRIPTION
**Pull request by alevy**
I've used what seems to be the rough consensus from #528 to update the system call interface document with proposed system call numbers. [rendered here](https://github.com/alevy/tock/blob/tock1/syscalls/doc/Syscalls.md#allocated-driver-numbers) for your convenience.

When we have agreement, but before merging, we'll do the following:

- [x] Update system call numbers in the code, both userland and kernel
- [x] Settle on a system call interface for drivers to be included in Tock 1.0
  - [x] Timer
  - [x] Debug Console
  - [x] LEDs
  - [x] Buttons
  - [x] ~~GPIO~~
  - [x] ADC
  - [x] Light sensor
  - [x] Temperature
  - [x] Humidity
- [x] Docuemt the system call interfaces in kernel and userspace
  - [x] Timer
  - [x] Debug Console
  - [x] LEDs
  - [x] Buttons
  - [x] ~~GPIO~~
  - [x] ADC
  - [x] Light sensor
    - [x] Update reference to driver number in tutorial 04
  - [x] Temperature
  - [x] Humidity
- [x] Implement the system call interfaces in kernel and userspace
  - [x] Timer
  - [x] Debug Console
  - [x] LEDs
  - [x] Buttons
  - [x] ~~GPIO~~
  - [x] ADC
  - [x] Light sensor
  - [x] Temperature
  - [x] Humidity

### CODE REVIEW

**Comment by bradjc** [comment]
What are your thoughts on something like this? (putting this driver number in the capsule itself)

```diff
git diff
diff --git a/boards/hail/src/main.rs b/boards/hail/src/main.rs
index 391cb8b6..a04cb5c1 100644
--- a/boards/hail/src/main.rs
+++ b/boards/hail/src/main.rs
@@ -77,7 +77,7 @@ impl Platform for Hail {
     {
 
         match driver_num {
-            0 => f(Some(self.console)),
+            console::DRIVER_NUM => f(Some(self.console)),
             1 => f(Some(self.gpio)),
 
             3 => f(Some(self.timer)),
diff --git a/capsules/src/console.rs b/capsules/src/console.rs
index 518f7ba7..d59f3fcc 100644
--- a/capsules/src/console.rs
+++ b/capsules/src/console.rs
@@ -42,6 +42,8 @@ use kernel::common::take_cell::TakeCell;
 use kernel::hil::uart::{self, UART, Client};
 use kernel::process::Error;
 
+pub const DRIVER_NUM: usize = 0x00000;
+
 pub struct App {
     write_callback: Option<Callback>,
     read_buffer: Option<AppSlice<Shared, u8>>,
```

It would make it a little easier to get right for new boards, and puts it in the source rather in just the documentation (and libtock). The downside is it obscures things a little bit more.

**Comment by bradjc** [comment]
Also, we're going to have a growing number of capsules. Is there a way we could use directories to separate the different types of capsules? In particular, I'm thinking right now of capsules that implement `Driver` and those that do not as being separate. I've kinda used the `_driver.rs` naming convention to indicate this so far.

I think the for the most part this would work well. One issue, however is capsules with both an internal "kernel" api and a separate `Driver` interface. These are chips that have an optional syscall interface, but could also be used by the kernel directly (depending on the needs of the platform). The [LTC294x](https://github.com/helena-project/tock/blob/master/capsules/src/ltc294x.rs) is implemented this way.

There also may be other distinctions (virtual_*.rs capsules, IC drivers, useful libraries, etc.) but I don't think we need to consider all of them now.

**Comment by alevy** [comment]
> What are your thoughts on something like this? (putting this driver number in the capsule itself)

I think this is a good idea. Given that we're using a `match` statement, the compiler ought to catch issues where driver numbers overlap.

**Comment by alevy** [comment]
> Also, we're going to have a growing number of capsules. [...]
I definitely agree this is a problem. Though, I don't think it's important for 1.0, so don't think we should worry about getting it 100% right.

For the moment, I think it makes sense to maybe have a few submodules: `capsules/src/mux` for virtualization capsules, `capsules/src/syscall` for system call drivers, `capsules/src/peripherals` for peripheral drivers. There might be some overlap, in which case we can temporarily make a tie-breaking decision ad-hoc, since it's just an organizational question anyway.

File: userland/libtock/led.c [hunk 1]
```diff
@@ -1,18 +1,20 @@
-#include "led.h"
-#include "tock.h"
+#include "internal/led.h"
+
+// For LEDs, these simply pass through to the syscall wrappers (the functions
+// in the internal folder).
```
Comment (brghena) [inline userland/libtock/led.c]: There's something I'm not understanding. Why did you move the syscall wrappers to the `internal` folder? It's certainly better commented and has the syscall number set up correctly, but why didn't those just go into `led.[c|h]`?

Comment (alevy) [inline userland/libtock/led.c]: I also agree that I don't understand why exactly. @bradjc how do you feel about us merging this back into the non-internal files?

Comment (bradjc) [inline userland/libtock/led.c]: I was under the impression that was the structure you wanted to move to. I can't remember exactly where it was said, or if it was said explicitly, but it started in #419. Why do it for only timers? How does one know when to create an internal header file?

Comment (alevy) [inline userland/libtock/led.c]: So, timers basically have _two_ interfaces. One of them (the internal one) actually talks directly to the kernel timers. You would use this one if you wanted to bypass virtual timers, or implement your own, say to hook into a different language. The other actually implements virtual timers. The internal API is not a "supported API" because if you use it it'll screw up the virtual timers. So it's in internal cause you really shouldn't use it, but it's there.

**Comment by alevy** [comment]
I'm fixing some bugs still, but currently my plan is to break up the 9dof HIL into three separate traits and (more importantly for this PR) have a separate system call interface for each the gyro, accelerometer and magnetometer. Thoughts? @bradjc in particular, are you offended by that? In the kernel, this ought to move towards your vision of a single virtualization implementation for all sensors.

**Comment by bradjc** [comment]
I added the docs for LEDs and buttons. I also updated buttons to allow us to define if a 1 or a 0 in the callback means the button is pressed or not pressed (which depends on the hardware).

**Comment by alevy** [comment]
Only ADC and GPIO left, from @brghena & @ppannuto

**Comment by bradjc** [comment]
I think we aren't doing a good job of documenting `allow`. From

https://github.com/helena-project/tock/blob/277285df88aef4a804dabbe4eba986bfd1cf34d4/kernel/src/sched.rs#L111-L128

I think that allow can always return `EINVAL` even if the capsule driver doesn't return it, and even if `allow` isn't even implemented in the capsule.

**Comment by alevy** [comment]
OK, done!

**Comment by ppannuto** [comment]
Do we want to do something about the `pin_config` mechanism:

https://github.com/helena-project/tock/blob/4ed764ffd399dcc3d7d344fd7cadf108362ad960/capsules/src/gpio.rs#L163
https://github.com/helena-project/tock/blob/4ed764ffd399dcc3d7d344fd7cadf108362ad960/capsules/src/gpio.rs#L189

I don't necessarily have a better solution, and I actually think the current approach is pretty okay. I pushed a commit that more clearly documents the usage. The only thing I might change is the values for `pin_config` (weird to me that `0` is active and `2` is inactive).

**Comment by alevy** [comment]
@ppannuto if you're gonna make that change, can you update the documentation I wrote in `docs/syscalls/0004_gpio.md`. It's not inconsistent with the implementation.

Re: `pin_config` both 5 and 7 set the pin_config, and they serve different roles. One turns the GPIO into an input and the other enables interrupts. If we get rid of 5, there is no way to read a pin's value without turning on interrupts.

**Comment by bradjc** [comment]
You also need to change gpio_async to match

https://github.com/alevy/tock/blob/dd7a92a8e907d649bc1f7aa491861d14c7b76b53/capsules/src/gpio_async.rs#L47-L56

which shares that enum in userland.

**Comment by ppannuto** [comment]
@alevy Oh duh, sorry. I'd imagined a `interrupt_config` => 0 => no interrupts for 7 when I asked that, but not really worth it.

@brghena Do you still feel that it's "clunky"? I'd rather not have a comment marking our 1.0 interface as clunky floating around.... :)

**Comment by brghena** [comment]
No, it's a pretty standard operation now. ADC does it, for instance. Please delete that comment.

File: doc/syscalls/60000_ambient_temperature.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+---
+driver number: 0x60000
+---
+
+# Humidity
```
Comment (bradjc) [inline doc/syscalls/60000_ambient_temperature.md]: There's some weirdness between these two files...the text doesn't match the filename.

**Comment by ppannuto** [comment]
I don't think there's a great way to make a PR against a PR.., so I just committed 7273e87, as I'm pretty sure it's the right choice. That said, please give it a a look and let me know if you disagree / I missed something.

**Comment by bradjc** [comment]
Can we ever imagine a platform with 257 GPIO pins? Maybe we should make the pin selection bits 16 bits.

Re 7273e87, why not go the other way and combine more of the calls? For instance, why not just have number 2 work (i.e. set the pin to output), instead of "Set the output of a GPIO pin (high). Using this command without first enabling output is undefined." Is there a subtle hardware thing we are working around, or could we just make this easy for users?

**Comment by alevy** [comment]
> Is there a subtle hardware thing we are working around, or could we just make this easy for users?

In general there are chips where you can choose either output _or_ input, but not both.

**Comment by alevy** [comment]
> Can we ever imagine a platform with 257 GPIO pins

I guess the question is, could we use the remaining 8-bits better in the future? I think the answer is no, and 16k pins seems like a safer future-proof interface than 256.

**Comment by ppannuto** [comment]
Re 16 bits, sounds good. Should we also extend gpio_async? Currently:
>    /// `data` is a 32 bit value packed with the lowest 8 bits as the port
>    /// number, the next lowest 8 bits as the pin number, and the remaining
>    /// upper bits as a command-specific value.

It almost feels more likely to me that there'd be a giant gpio extender before a chip with >256 gpios, but who knows, both feel a bit fanciful right now

----

Re combining calls, there's some deeper issues there:
> In general there are chips where you can choose either output or input, but not both.

Indeed, currently the GPIO HIL doesn't allow `make_output` to ever fail (or any of the complimentary functions), but that probably needs to be changed (or create `OutputPin` and `InputPin` types, but that's just moving the issue around)

**Comment by bradjc** [comment]
I just added the 4th argument (register r3) to the `command` call. We can now update any interfaces that are currently using bitpacking in command arguments that could benefit from using the fourth argument.

**Comment by alevy** [comment]
GPIO, GPIO Async and ADC interfaces updated to use the second argument in command instead of bitpacking
